File: mysite/simulation/alt/robot.py
import math
from math import pi


class Box:
	def __init__(self, width=0, length=0):
		self._length = length
		self._width = width

	@property
	def length(self):
		return self._length

	@length.setter
	def length(self, val):
		self._length = val

	@property
	def width(self):
		return self._width

	@width.setter
	def width(self, val):
		self._width = val


class Position:
	def __init__(self, x=0, y=0):
		self._x = x
		self._y = y

	@property
	def x(self):
		return self._x

	@x.setter
	def x(self, val):
		self._x = val

	@property
	def y(self):
		return self._y

	@y.setter
	def y(self, val):
		self._y = val

	@property
	def pos(self):
		return Position(self._x, self._y)

	@pos.setter
	def pos(self, val):
		self._x = val.x
		self._y = val.y


class State:
	def __init__(self, x=0, y=0, theta=0, vel=0):
		self._x = x
		self._y = y
		self._theta = theta
		self._vel = vel

	@property
	def x(self):
		return self._x

	@x.setter
	def x(self, val):
		self._x = val

	@property
	def y(self):
		return self._y

	@y.setter
	def y(self, val):
		self._y = val

	@property
	def theta(self):
		return self._theta

	@theta.setter
	def theta(self, val):
		self._theta = val

	@property
	def pos(self):
		return Position(self._x, self._y)

	@pos.setter
	def pos(self, val):
		self._x = val.x
		self._y = val.y

	@property
	def vel(self):
		return self._vel

	@vel.setter
	def vel(self, val):
		self._vel = val


class Robot:
	# class variables
	WIDTH = 9.0
	LENGTH = 10.5
	AXLE_LENGTH = 8.4
	WHEEL_DIAMETER = 5.0
	WHEEL_CIRCUMFERENCE = 15.3938
	IMU = Position(0.0, 0.0)
	FRONT_LASER = Position(0.0, 2.5)
	RIGHT_LASER = Position(3.0, -2.5)
	LEFT_WHEEL = Position(-4.2, 0.0)
	RIGHT_WHEEL = Position(4.2, 0.0)
	FRONT_LEFT = Position(-4.2, 3.7)
	FRONT_RIGHT = Position(4.2, 3.7)
	BACK_LEFT = Position(-4.2, -7.1)
	BACK_RIGHT = Position(4.2, -7.1)
	LASER_MAX_RANGE = 150.0
	MAX_RPM = 100
	MAX_CM_PER_MS = (MAX_RPM/60000) * WHEEL_CIRCUMFERENCE
	ANALOG_MAX = 255

	# Rotates position about axis by angle counterclockwise.
	def rotate(rel_pos, d_theta):
		cos = math.cos(d_theta)
		sin = math.sin(d_theta)
		return Position(
			rel_pos.x * cos - rel_pos.y * sin,
			rel_pos.x * sin + rel_pos.y * cos
		)

	# Translates a position in field frame.
	def translate(pos, dx, dy):
		return Position(
			pos.x + dx,
			pos.y + dy
		)

	# Rotates and translates in field frame.
	def transform(rel_pos, d_state):
		pos = Robot.rotate(rel_pos, d_state.theta)
		pos = Robot.translate(pos, d_state.x, d_state.y)
		return pos

	# Keep angle between -pi and pi
	def reduce_angle(theta):
		return (theta % (2*pi)) - (2*pi)


	# Quadrant of angle
	def quadrant(theta):
		quad = int(round(Robot.reduce_angle(theta) / (pi/4))) + 2
		quad_map = [4, 1, 2, 3]
		return quad_map[quad]


	# Distance from a point in one direction
	# [distance to wall, wall projected on, point on wall projected on]
	def distance_from_point(state):
		eps = 0.1 # eliminate large calculations
		cos = math.cos(state.theta)
		sin = math.sin(state.theta)
		abs_theta = abs(state.theta)

		# walls that the robot might be facing
		vert_wall = Box.length if state.theta < 0 else 0
		hori_wall = Box.width if abs_theta < pi/2 else 0

		# distance to each relevant wall
		if abs(sin) > eps:
			vert_dist = abs(vert_wall - state.x) / sin
		else:
			vert_dist = math.inf

		if abs(cos) > eps:
			hori_dist = abs(hori_wall - state.y) / cos
		else:
			hori_dist = math.inf

		quad = Robot.quadrant(state.theta)
		
		# laser projects on side walls
		if vert_dist < hori_dist:
			min_dist = vert_dist
			nearest_wall = "left" if quad == 2 or quad == 3 \
				else "right"
			proj_pos = state.y + min_dist * cos

		# laser projects on top or bottom wall
		else:
			min_dist = hori_dist
			nearest_wall = "top" if quad == 1 or quadrant == 2 \
				else "bottom"
			proj_pos = state.x - min_dist * sin

		return [min_dist, nearest_wall, proj_pos]


	def __init__(self, startState=None, isPrev=False):
		
		# state of robot
		self._state = startState or State()
		self._period = 5.0
		self._loop = None
		self._task = None

		# parts of the robot
		self._imu = {"mag": 0, "gyro": 0, "accel": 0}
		self._front_laser = {"dist": 0}
		self._right_laser = {"dist": 0}
		self._left_wheel = {}
		self._right_wheel = {}
		self._front_left = {}
		self._front_right = {}
		self._back_left = {}
		self._back_right = {}

		# save positions relative to field
		self.update_positions()

		# input
		self._left_signal = False
		self._right_signal = False

		# previous state
		if isPrev == False:
			self._prev = Robot(self._state, True)


	def update_positions(self):
		self._imu['pos'] = Robot.transform(Robot.IMU, self._state)
		self._front_laser['pos'] = Robot.transform(Robot.FRONT_LASER, self._state)
		self._right_laser['pos'] = Robot.transform(Robot.RIGHT_LASER, self._state)
		self._left_wheel['pos'] = Robot.transform(Robot.LEFT_WHEEL, self._state)
		self._right_wheel['pos'] = Robot.transform(Robot.RIGHT_WHEEL, self._state)
		self._front_left['pos'] = Robot.transform(Robot.FRONT_LEFT, self._state)
		self._front_right['pos'] = Robot.transform(Robot.FRONT_RIGHT, self._state)
		self._back_left['pos'] = Robot.transform(Robot.BACK_LEFT, self._state)
		self._back_right['pos'] = Robot.transform(Robot.BACK_RIGHT, self._state)

File: js/robot.js
class Position {
	constructor(x=0, y=0) {
		this.x = x;
		this.y = y;
	}

	get pos() {
		return this;
	}

	set pos(position) {
		this.x = position.x;
		this.y = position.y;
	}
}


class State {
	constructor(x=0, y=0, theta=0, v=0) {
		this.update(x, y, theta, v);
	}

	get pos() {
		return new Position(this.x, this.y);
	}

	set pos(position) {
		this.x = position.x;
		this.y = position.y;
	}

	update(x, y, theta, v=0) {
		this.x = x;
		this.y = y;
		this.theta = theta;
		this.v = v;
	}
}


class Robot {

	static get WIDTH() { return 9.0; }
	static get LENGTH() { return 10.5; }
	static get AXLE_LENGTH() { return 8.4; }
	static get WHEEL_DIAMETER() { return 5.0; }
	static get WHEEL_CIRCUMFERENCE() { return 15.3938; }
	static get IMU() { return new Position(0.0, 0.0); }
	static get FRONT_LASER() { return new Position(0.0, 2.5); }
	static get RIGHT_LASER() { return new Position(3.0, -2.5); }
	static get LEFT_WHEEL() { return new Position(-4.2, 0.0); }
	static get RIGHT_WHEEL() { return new Position(4.2, 0.0); }
	static get FRONT_LEFT() { return new Position(-4.2, 3.7); }
	static get FRONT_RIGHT() { return new Position(4.2, 3.7); }
	static get BACK_LEFT() { return new Position(-4.2, -7.1); }
	static get BACK_RIGHT() { return new Position(4.2, -7.1); }
	static get LASER_MAX_RANGE() { return 150.0; }
	static get MAX_RPM() { return 100.0; }
	static get MAX_CM_PER_MS() { return (Robot.MAX_RPM/60000) * Robot.WHEEL_CIRCUMFERENCE; }
	static get ANALOG_MAX() { return 255; }

	
	static rotate(rel_pos, dtheta) {
		var cos = Math.cos(dtheta);
		var sin = Math.sin(dtheta);

		return new Position(
			rel_pos.x * cos - rel_pos.y * sin,
			rel_pos.x * sin + rel_pos.y * cos
		);
	}


	static translate(pos, dx, dy) {
		return new Position(
			pos.x + dx,
			pos.y + dy
		);
	}


	static transform(rel_pos, dstate) {
		return Robot.translate(
			Robot.rotate(rel_pos, dstate.theta), 
			dstate.x, 
			dstate.y
		);
	}


	static next_orientation(state, arc_length, radius, dtheta) {
		var cos = Math.cos(state.theta);
		var sin = Math.sin(state.theta);
		var cos_delta = Math.cos(state.theta + dtheta);
		var sin_delta = Math.sin(state.theta + dtheta);

		var x;
		var y;

		if (radius < 200) {
			x = radius*(cos_delta - cos) + state.x;
			y = radius*(sin_delta - sin) + state.y;
		} else {
			x = state.x - arc_length * sin;
			y = state.y + arc_length * cos;
		}

		return new State(x, y, Robot.reduce_angle(state.theta + dtheta));
		
	}


	static reduce_angle(theta) {
		var angle = theta % (2*Math.PI);
		if (angle < -Math.PI) {
			angle += 2*Math.PI;
		} else if (angle > Math.PI) {
			angle -= 2*Math.PI;
		}
		return angle;
	}


	static quadrant(theta) {
		var quad = Math.floor(Robot.reduce_angle(theta)/(Math.PI/4)) + 2;
		var quad_map = [4, 1, 2, 3];
		return quad_map[quad];
	}


	static distance_from_point(state, box) {
		const EPS = 0.1;
		var cos = Math.cos(state.theta);
		var sin = Math.sin(state.theta);

		var vert_wall = (state.theta < 0 ? box.length_cm : 0);
		var hori_wall = (Math.abs(state.theta) < Math.PI/2 ? box.width_cm : 0);

		var vert_dist;
		var hori_dist;

		if (Math.abs(sin) > EPS) {
			vert_dist = Math.abs((vert_wall - state.x) / sin);
		} else {
			vert_dist = Infinity;
		}

		if (Math.abs(cos) > EPS) {
			hori_dist = Math.abs((hori_wall - state.y) / cos);
		} else {
			hori_dist = Infinity;
		}
		
		var quad = Robot.quadrant(state.theta);
		var min_dist;
		var nearest_wall;
		var proj_pos;

		if (vert_dist < hori_dist) {
			min_dist = vert_dist;
			nearest_wall = (quad == 2 || quad == 3 ? "left" : "right");
			proj_pos = state.y + min_dist * cos;
		} else {
			min_dist = hori_dist;
			nearest_wall = (quad == 1 || quad == 2 ? "top" : "bottom");
			proj_pos = state.x - min_dist * sin;
		}

		return { dist: min_dist, 
						 wall: nearest_wall, 
						 pos: proj_pos };

	}


	static random_angle() {
		return Math.random() * 2.0 * Math.PI - Math.PI;
	}


	static random_float(min, max) {
		return Math.random() * (max - min) + min;
	}


	static gaussian(mean=0, variance=1) {
		// central limit theorem
		var sum = 0;
		var n = Math.sqrt(12*variance);
		for (var i = 0; i < 10; i += 1) {

			sum += Math.random()*n - (n/2 - mean);

		}
		return sum/10.0;
	}


	constructor(startState=null, isPrev=false) {
		this.state = startState || new State();
		this.period = 5.0;
		
		this.imu = { mag: 0, gyro: 0, accel: 0 };
		this.front_laser = { dist: 0 };
		this.right_laser = { dist: 0 };
		this.left_wheel = {}
		this.right_wheel = {}
		this.front_left = {}
		this.front_right = {}
		this.back_left = {}
		this.back_right = {}

		this.update_positions();

		this.left_signal = false;
		this.right_signal = false;

		if (isPrev == false) {
			this.prev = new Robot(self.state, true);
		}
	}


	update_positions() {
		this.imu.pos = Robot.transform(Robot.IMU, this.state);
		this.front_laser.pos = Robot.transform(Robot.FRONT_LASER, this.state);
		this.right_laser.pos = Robot.transform(Robot.RIGHT_LASER, this.state);
		this.left_wheel.pos = Robot.transform(Robot.LEFT_WHEEL, this.state);
		this.right_wheel.pos = Robot.transform(Robot.RIGHT_WHEEL, this.state);
		this.front_left.pos = Robot.transform(Robot.FRONT_LEFT, this.state);
		this.front_right.pos = Robot.transform(Robot.FRONT_RIGHT, this.state);
		this.back_left.pos = Robot.transform(Robot.BACK_LEFT, this.state);
		this.back_right.pos = Robot.transform(Robot.BACK_RIGHT, this.state);
	}
}
